### PR TITLE
DEV: Implement Distributed mutex in lua

### PR DESCRIPTION
**Pull request by CvX**
The rationale behind this was mostly to stop using `redis.synchronize` (now removed in redis gem 4.6)

### CODE REVIEW

**Comment by tgxworld** [comment]
Are using a LUA script to ensure atomic execution? If so I'm wondering why we don't use `Redis#multi` or `Redis#exec` which ensures that the commands are ran in an atomic fashion as well.

File: spec/lib/distributed_mutex_spec.rb [hunk 1]
```diff
@@ -40,26 +40,19 @@
...
-  # ./lib/distributed_mutex.rb:33:in `block in synchronize'
-  xit 'allows the validity of the lock to be configured' do
+  it "allows the validity of the lock to be configured" do
     freeze_time
-
     mutex = DistributedMutex.new(key, validity: 2)
 
     mutex.synchronize do
-      expect(Discourse.redis.ttl(key)).to eq(2)
+      expect(Discourse.redis.ttl(key)).to eq(3)
```
Comment (davidtaylorhq) [inline spec/lib/distributed_mutex_spec.rb:48]: This feels kinda risky... `freeze_time` is not gonna affect Redis's internal clock. So if it takes a little longer for this line to be run, the ttl could be lower? Maybe that's why this test was skipped?

Maybe we should be more generous, like `.to be <= 3` ? 🤔

(and similarly on the other ttl check a few lines lower)

**Comment by davidtaylorhq** [comment]
> I'm wondering why we don't use Redis#multi or Redis#exec which ensures that the commands are ran in an atomic fashion as well

I was wondering the same - looks like `Redis#multi/exec` allows you to run a sequence of operations atomically, but those operations cannot depend on the result of a previous operation. They all have to be totally independent.

You can have some level of interdependency using `Redis#watch` using the [check and set](https://redis.io/docs/manual/transactions/#optimistic-locking-using-check-and-set) approach, which is what we were doing previously. But that is problematic for us because our redis connection is shared between threads, and so we have to wrap the whole `watch` -> `multi` in a connection-level mutex. That connection-level mutex has been made private in the latest Redis client gem, hence the need for this change.

Using LUA kills two birds with one stone - it takes care of the redis-level atomicity, and it means things happen within a single call so there's no need for a connection-level mutex in Ruby. Also, the official guide [on transactions](https://redis.io/docs/manual/transactions/#optimistic-locking-using-check-and-set) says:

>  Everything you can do with a Redis Transaction, you can also do with a script, and usually the script will be both simpler and faster.

**Comment by tgxworld** [comment]
> Also, the official guide [on transactions](https://redis.io/docs/manual/transactions/#optimistic-locking-using-check-and-set) says:

> Everything you can do with a Redis Transaction, you can also do with a script, and usually the script will be both simpler and faster.

Come to think of it, I actually found this to be true. The LUA script was way easier to read and understand than the Ruby code which we had. I just had to do a crash course on LUA in order to understand it. Nothing too difficult though.

File: lib/distributed_mutex.rb [hunk 2]
```diff
@@ -50,79 +77,32 @@ def synchronize
...
-
   def get_lock
     attempts = 0
 
     while true
-      got_lock, expire_time = try_to_get_lock
-      if got_lock
-        return expire_time
-      end
+      got_lock, expire_time = LOCK_SCRIPT.eval(redis, [prefixed_key], [validity])
```
[the remark this answers is not in the document]
Comment (CvX) [reply lib/distributed_mutex.rb:84]: Did you mean this? 🙂 

```diff
--- a/lib/distributed_mutex.rb
+++ b/lib/distributed_mutex.rb
@@ -12,7 +12,7 @@ class DistributedMutex
     local current_expire_time = redis.call("get", KEYS[1])
 
     if current_expire_time and tonumber(now) <= tonumber(current_expire_time) then
-      return { false, expire_time }
+      return { false, nil }
     else
       local result = redis.call("setex", KEYS[1], ARGV[1] + 1, tostring(expire_time))
       return { true, expire_time }
```

File: spec/lib/distributed_mutex_spec.rb (hunk 1 above)
Comment (CvX) [reply spec/lib/distributed_mutex_spec.rb]: Updated the spec. What do you think? 🙂

Comment (davidtaylorhq) [reply spec/lib/distributed_mutex_spec.rb]: Looks great to me 🚀

**Review by davidtaylorhq — APPROVED** [review]

File: lib/distributed_mutex.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgxworld) [reply lib/distributed_mutex.rb:84]:
```diff
--- a/lib/distributed_mutex.rb
+++ b/lib/distributed_mutex.rb
@@ -12,7 +12,7 @@ class DistributedMutex
     local current_expire_time = redis.call("get", KEYS[1])
 
     if current_expire_time and tonumber(now) <= tonumber(current_expire_time) then
-      return { false, expire_time }
+     return nil
     else
       local result = redis.call("setex", KEYS[1], ARGV[1] + 1, tostring(expire_time))
-       return { true, expire_time }
+      return expire_time
```

Something like that.

**Pull request merged**